### PR TITLE
Add last update

### PR DESCRIPTION
**Pull request by y0z**
## Motivation

Displays the most recently updated packages first.

## Description of the changes

- Add `updated_at` property to the index.md when processing the data of optunahub-registry based on the Git information.
- Order the packages based on `updated_at`.
- Display the last update information.

### Top page
image

### Package detail page
image

### CODE REVIEW

**Comment by ofk** [comment]
LGTM as a web frontend-er.

File: .github/workflows/gh-pages.yml [hunk 1]
```diff
@@ -19,8 +19,6 @@ jobs:
     steps:
       - name:
         uses: actions/checkout@v4
-        with:
-          ref: main
```
Comment (toshihikoyanase) [inline .github/workflows/gh-pages.yml:23]: Why did you remove this condition?

File: hugo_build.py [hunk 2]
```diff
@@ -1,20 +1,47 @@
 import os
 import shutil
 
+import datetime
```
Comment (toshihikoyanase) [inline hugo_build.py:4]: A nit.
```suggestion
import datetime
import os
import shutil
```

File: hugo_build.py [hunk 3]
```diff
@@ -1,20 +1,47 @@
...
+
+    package_dir = "package"
+    registry_dir = f"optunahub-registry/{package_dir}"
     categories = ["samplers", "pruners", "visualization"]
     for c in categories:
         packages = os.listdir(f"{registry_dir}/{c}")
         for p in packages:
             hugo_dir = f"content/{c}/{p}"
+
+            # Copy README.md and images
```
Comment (toshihikoyanase) [inline hugo_build.py:23]: a nit.
```suggestion
            # Copy README.md and images.
```

File: hugo_build.py [hunk 4]
```diff
@@ -1,20 +1,47 @@
...
             os.makedirs(hugo_dir, exist_ok=True)
-            shutil.copy(f"{registry_dir}/{c}/{p}/README.md", f"{hugo_dir}/index.md")
             imgdir = f"{registry_dir}/{c}/{p}/images"
             if os.path.exists(imgdir):
                 shutil.copytree(imgdir, f"{hugo_dir}/images", dirs_exist_ok=True)
 
+            # Read README.md and create index.md from it with modifications
```
Comment (toshihikoyanase) [inline hugo_build.py:29]: ditto.
```suggestion
            # Read README.md and create index.md from it with modifications.
```

File: hugo_build.py [hunk 5]
```diff
@@ -1,20 +1,47 @@
...
+            with open(f"{registry_dir}/{c}/{p}/README.md", "r") as readme_md:
+                contents = frontmatter.load(readme_md)
+                with open(f"{hugo_dir}/index.md", "w") as index_md:
+                    # Last update for the package
```
Comment (toshihikoyanase) [inline hugo_build.py:33]: ditto.
```suggestion
                    # Last update for the package.
```

File: hugo_build.py [hunk 6]
```diff
@@ -1,20 +1,47 @@
...
+                    updated_at_epoch = next(
+                        repo.iter_commits(
+                            paths=tree[f"{package_dir}/{c}/{p}"].path, max_count=1
+                        )
+                    ).committed_date
+                    updated_at_iso = datetime.datetime.fromtimestamp(
+                        updated_at_epoch
+                    ).isoformat()
```
Comment (toshihikoyanase) [inline hugo_build.py:41]: As can be seen on the top page, how about showing the date?
The current implementation might be too specific, and users might get confused due to the difference in time zones.

```suggestion
                    updated_at_iso = datetime.datetime.fromtimestamp(
                        updated_at_epoch
                    ).date().isoformat()
```

**Review by toshihikoyanase — COMMENTED** [review]
Thanks for your PR. I added some comments.

File: .github/workflows/gh-pages.yml (hunk 1 above)
Comment (y0z) [reply .github/workflows/gh-pages.yml:23]: This forces CI to always check out the main branch instead of the branch for this PR.
This results in "Error: requirements.txt is not found" because the file has been introduced by this PR.

Comment (toshihikoyanase) [reply .github/workflows/gh-pages.yml:23]: That makes sense. We should check out the feature branch for pull-request events.

File: hugo_build.py (hunk 6 above)
Comment (y0z) [reply hugo_build.py:40]: Information on not only the date but also the time is used to sort packages properly.

So, instead of removing the time information, I want to apply formatting to display only the date information in `single.html`.

image

File: themes/hub/layouts/_default/single.html [hunk 7]
```diff
@@ -28,6 +28,8 @@
         {{- end }}
       </ul>
     </dd>
+    <dt class="float-start me-2">Last update</dt>
+    <dd class="clearfix mb-1">{{ time.Format "2006-01-02" (time.AsTime .Params.updated_at) }}</dd>
```
Comment (y0z) [inline themes/hub/layouts/_default/single.html:32]: FYI, the date "2006-01-02" is usually used in HUGO for formatting.
https://gohugo.io/methods/time/format/#examples

File: hugo_build.py (hunk 6 above)
Comment (toshihikoyanase) [reply hugo_build.py:40]: Thanks. That must be better.

**Review by toshihikoyanase — APPROVED** [review]
It worked as expected. LGTM!

image

image
